### PR TITLE
feat: scheduler (9/): add uniquename utility for preparing binding names

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR is part of the PRs that implement the Fleet workload scheduling.

It features a utility that generates unique names for resource bindings.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

- [x] Unit tests

### Special notes for your reviewer

Per earlier agreement, unit tests in this PR have a limited scope.

### CODE REVIEW

File: pkg/scheduler/framework/uniquename/uniquename.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+// * [CRP-NAME] - [TARGET-CLUSTER-NAME] - [RANDOM-SUFFIX]
+//
+// Segments will be truncated if necessary.
+//
+// Note that the name generation is, in essence, a best-effort process, though the chances
+// of name collisions are extremely low.
+//
+// In addition, note that this function assumes that both the CRP name and the cluster name
+// are valid DNS subdomain names (RFC 1123).
+func ClusterResourceBindingUniqueName(CRPName string, clusterName string) (string, error) {
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/uniquename/uniquename.go:42]:
```suggestion
func NewClusterResourceBindingName(CRPName string, clusterName string) (string, error) {
```
The caller will call this func like "uniquename.ClusterResourceBindingUniqueName" and it has duplicate info there.

File: pkg/scheduler/framework/uniquename/uniquename.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+	slotsPerSeg := (validation.DNS1123SubdomainMaxLength - reservedSlots) / 2
+	uniqueName := fmt.Sprintf("%s-%s-%s",
+		CRPName[:minInt(slotsPerSeg, len(CRPName))],
+		clusterName[:minInt(slotsPerSeg, len(clusterName))],
+		uuid.NewUUID()[:uuidLength],
+	)
+
+	if errs := validation.IsDNS1123Subdomain(uniqueName); len(errs) != 0 {
+		return "", fmt.Errorf("failed to format a unique RFC 1123 DNS subdomain name with namespace %s, name %s: %v", CRPName, clusterName, errs)
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/uniquename/uniquename.go:53]: just would like to clarify that, it should never happen in the production.

Comment (michaelawyu) [reply pkg/scheduler/framework/uniquename/uniquename.go:54]: Added a comment.

File: pkg/scheduler/framework/uniquename/uniquename.go [hunk 3]
```diff
@@ -26,7 +26,7 @@ func minInt(a, b int) int {
 	return b
 }
 
-// ClusterResourceBindingUniqueName returns a unique name for a cluster resource binding in the
+// NewlusterResourceBindingName returns a unique name for a cluster resource binding in the
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/uniquename/uniquename.go:29]: should be "NewClusterResourceBindingName"

missing "C" :)

Comment (michaelawyu) [reply pkg/scheduler/framework/uniquename/uniquename.go:29]: Oh, my bad. Adding it now.

**Review by zhiying-lin — APPROVED** [review]

**Pull request merged**